File: piper/extensions/chrome/src/engine/statements/math_number_property.js
import { stringify } from "../util"


function isPrime(num) {
    for (let i = 2; i < num; i++) {
        if (num % i === 0) {
            return false
        }
    }
    return num > 1
}

export default async function(statement, onError) {
    try {
        let pointer = statement.pointer
        let value1 = await this.executeStatement(stringify(pointer.value1))
        let prop = stringify(pointer.property)

        switch(prop) {
            case "EVEN":
                return value1 % 2 === 0
            case "ODD":
                return value1 % 2 !== 0
            case "PRIME":
                return isPrime(value1)
            case "WHOLE":
                return value1 % 1 === 0
            case "POSITIVE":
                return value1 > 0
            case "NEGATIVE":
                return value1 < 0
            case "DIVISIBLE_BY":
                let value2 = await this.executeStatement(stringify(pointer.value2))
                return value1 % value2 === 0
        }
    } catch (e) {
        onError(e)
    }

    return false
}

File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/math_number_property.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import io.gopiper.piper.util.isPrime
import org.json.JSONObject
import java.lang.Exception

suspend fun Piper.executeMathNumberProperty(statement: JSONObject, onError: (e: Exception) -> Unit): Boolean {
    try {
        Log.d("hello", "execute_math_number_property")
        val pointer = statement.getJSONObject("pointer")
        val value1 = executeStatement(pointer["value1"].toString())
        val prop = pointer["property"].toString()

        val value1Double = value1.toString().toDouble()

        return when (prop) {
            "EVEN" -> value1Double % 2.0 == 0.0
            "ODD" -> value1Double % 2.0 != 0.0
            "PRIME" -> if (value1 is Double) false else isPrime(value1.toString().toInt())
            "WHOLE" -> value1Double % 1 == 0.0
            "POSITIVE" -> value1Double > 0
            "NEGATIVE" -> value1Double < 0
            "DIVISIBLE_BY" -> value1Double % executeStatement(pointer["value2"].toString()).toString().toInt() == 0.0
            else -> false
        }
    } catch (e: Exception) {
        onError(e)
        return false
    }
}